packages/shell-api/src/shell-internal-state.ts

```diff
@@ -0,0 +1,238 @@
+import {
+  AggregationCursor,
+  CursorIterationResult,
+  Cursor,
+  Help,
+  Database,
+  Mongo,
+  ReplicaSet,
+  Shard,
+  signatures,
+  ShellBson,
+  toIterator,
+} from './index';
+import { EventEmitter } from 'events';
+import { DatabaseOptions, Document, ServiceProvider, ReplPlatform } from '@mongosh/service-provider-core';
+import { MongoshInvalidInputError, MongoshUnimplementedError } from '@mongosh/errors';
+import AsyncWriter from '@mongosh/async-rewriter';
+
+/**
+ * Anything to do with the internal shell state is stored here.
+ */
+export default class ShellInternalState {
+  public currentCursor: Cursor | AggregationCursor;
+  public currentDb: Database;
+  public messageBus: EventEmitter;
+  public asyncWriter: AsyncWriter;
+  public initialServiceProvider: ServiceProvider; // the initial service provider
+  public uri: string;
+  public connectionInfo: any;
+  constructor(initialServiceProvider: ServiceProvider, messageBus: any = new EventEmitter()) {
+    this.initialServiceProvider = initialServiceProvider;
+    this.messageBus = messageBus;
+    this.asyncWriter = new AsyncWriter(signatures);
+    const mongo = new Mongo(this);
+    this.currentCursor = null;
+    this.currentDb = mongo.getDB('test'); // TODO: set to CLI arg
+  }
+
+  async fetchConnectionInfo(): Promise<void> {
+    this.connectionInfo = await this.currentDb.mongo.serviceProvider.getConnectionInfo();
+    this.messageBus.emit('mongosh:connect', this.connectionInfo.extraInfo);
+  }
+
+  // TODO: do we need to close all Mongos ever made?
```

Comment:
Seems a nice thing to be doing.

Reply:
I've updated it so that we store all newly created mongos and close them all at the end.